src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -1183,12 +1183,23 @@ void LpNormCompute(const nnvm::NodeAttrs& attrs,
   } else {
     small = ReduceAxesShapeImpl(inputs[0].shape_, param.axis, true, false);
   }
+
   if (param.ord == 1) {
-    ReduceAxesComputeImpl<xpu, mshadow_op::sum, true, false, mshadow_op::abs>(
+    if (dmlc::GetEnv("MXNET_ENFORCE_SAFE_ACCUMULATION", false)) {
+      ReduceAxesComputeImpl<xpu, mshadow_op::sum, true, false, mshadow_op::abs>(
+        ctx, inputs, req, outputs, small);
+    } else {
+      ReduceAxesComputeImpl<xpu, mshadow_op::sum, false, false, mshadow_op::abs>(
         ctx, inputs, req, outputs, small);
+    }
   } else if (param.ord == 2) {
-    ReduceAxesComputeImpl<xpu, mshadow_op::nrm2, true, false, mshadow_op::identity>(
+    if (dmlc::GetEnv("MXNET_ENFORCE_SAFE_ACCUMULATION", false)) {
```

Comment:
Lets check if out_dtype is set and accordingly use safe accumulation as discussed in: https://github.com/apache/incubator-mxnet/issues/14760 . It would be better to not add another env variable, users can opt in by explicitly specifying out_dtype.

Reply:
There're 3 data types involved in an accumulation: input type, accumulation type and output type. The method you're suggesting is enforcing the latter 2 types to be the same. For a user that would like to use a lower precision, for example fp16, but still want a safe accumulation, `out_dtype` argument will actually lead to the requirement of an extra type cast after the accumulation, which, I think is even more burden than specifying an env var.

Reply:
Here is what I understand from the PRs: #14616 and #14098: when out_dtype is provided safe_accumulation is then set, to produce output of out_dtype and accumulation of next higher precision dtype (or accumulate in fp64 if out_dtype is fp64). For example, if out_dtype is provided as fp32, accumulation is done in fp64, if out_dtype is provided as uint8, accumulation is done in uint32. If my understanding is correct, if out_dtype is provided as fp16, accumulation will be done in fp32. Isn't this what you expect ? What is the burden here ?

Reply:
@anirudh2290 that may not always be true. For example, for the log_softmax operator with fp16 inputs, you could just have fp32 acc_type and fp32 out_type. Fp16 output does not make sense here because it always overflows.

Reply:
had a discussion with @eric-haibin-lin and @haojin2 . From the discussion, seems like users may not have use cases to configure individual ops with safe_acc. More often encountered use cases are global flag with all or none. Also, it should be possible to add this flag later without breaking backward compatibility if we find that customers need such use case. So I am okay with the current approach for using env variable with default to 0, since it doesnt break backward compatibility.